### PR TITLE
Add vector layer elevation properties

### LINKED ISSUE #7378: Add layer-level settings for vector/raster layer elevation properties (Request in QGIS)
### Request for documentation
From pull request QGIS/qgis#47522
Author: @nyalldawson
QGIS version: 3.26

**Add layer-level settings for vector/raster layer elevation properties**

### PR Description:
This PR partially implements https://github.com/qgis/QGIS-Enhancement-Proposals/issues/246 , by adding layer-level elevation properties for raster and vector layers.

These properties are set through a new "Elevation" tab in the raster/vector layer properties dialog.

## Raster elevation properties

image

## Vector elevation properties

image

(Raster elevation properties are currently unused, but vector elevation properties are used as the default properties which are inherited for new 3d symbols set for that layer)

### Commits tagged with [need-docs] or [FEATURE]

### LINKED ISSUE #7506: Allow for data defined vector feature extrusion and offset heights for profile charts (Request in QGIS)
### Request for documentation
From pull request QGIS/qgis#48177
Author: @nyalldawson
QGIS version: 3.26

**Allow for data defined vector feature extrusion and offset heights for profile charts**

### PR Description:
(Temporarily includes https://github.com/qgis/QGIS/pull/48165 )

Allows the extrusion height and feature elevation offset to be taken from a field in the layer, or based on an expression.

The data defined extrusion and offset values are respected in elevation profile charts, and will also be used as the default offset/extrusion settings for any new 3d polygon symbol created for that layer.

https://user-images.githubusercontent.com/1829991/162849365-b467a5ea-2571-43ee-ab9b-e5f132aaaf52.mp4

### Commits tagged with [need-docs] or [FEATURE]
"[feature] Allow for data defined vector feature extrusion and offset\nheights\n\nAllows the extrusion height and feature elevation offset to be\ntaken from a field in the layer, or based on an expression"

### PR DESCRIPTION
**Pull request by DelazJ**
fixes #7378, fixes #7506

### CODE REVIEW

File: docs/user_manual/working_with_vector/vector_properties.rst [hunk 1]
```diff
@@ -1999,12 +1996,25 @@ To display a layer in 3D, select from the combobox at the top of the tab, either
...
+.. attention:: **Prefer the** :guilabel:`Elevation` **tab for symbol elevation and terrain settings**
 
-.. figure:: img/3d_view_properties.png
-   :align: center
+ Features' elevation and altitude related properties (:guilabel:`Altitude clamping`,
+ :guilabel:`Altitude binding`, :guilabel:`Extrusion` or :guilabel:`Height`)
+ in the :guilabel:`3D View` tab inherit their default values from the layer's
+ :ref:`Elevation properties <vectorelevationmenu>` and should preferably be set
+ from within the :guilabel:`Elevation` tab.
+ They will be removed from the :guilabel:`3D View` tab in QGIS 4.0.
```
Comment (DelazJ) [inline docs/user_manual/working_with_vector/vector_properties.rst:2006]: Might be good to get confirmation for this

File: docs/user_manual/working_with_vector/vector_properties.rst [hunk 2]
```diff
@@ -3235,6 +3245,58 @@ More information on variables usage in the General Tools
...
+  footprints layers.
+* :guilabel:`Elevation Binding`: only relevant when combining an
+  :guilabel:`Elevation clamping` relying on the terrain with a line or
+  polygon layer, this option controls how feature elevation is set relative
+  to the terrain height. The terrain can be sampled:
+
+  * at the feature's :guilabel:`Centroid`, with the centroid height being
+    added to each vertex's z value
+  * at every individual :guilabel:`Vertex` before being added to the vertex's
+    z value
```
Comment (DelazJ) [inline docs/user_manual/working_with_vector/vector_properties.rst:3295]: I must admit that I do not understand the binding system and simply picked these sentences from the GUI. Still unclear. So if someone can explain/has a better wording?

**Review by DelazJ — COMMENTED** [review]
@nyalldawson may I get a review please? here and other [elevation/profile PR](https://github.com/qgis/QGIS-Documentation/pulls?q=is%3Apr+is%3Aopen+label%3A%22Profile+Tool%22)? Thanks

File: docs/user_manual/working_with_vector/vector_properties.rst (hunk 1 above)
Comment (nyalldawson) [reply docs/user_manual/working_with_vector/vector_properties.rst:2006]: I'd remove that line. It **should** happen, but it's not a trivial change and may not happen as a result.

File: docs/user_manual/working_with_vector/vector_properties.rst [hunk 3]
```diff
@@ -3235,6 +3245,58 @@ More information on variables usage in the General Tools
...
+
+
+* :guilabel:`Elevation Clamping`: defines how and whether the features altitude
+  should be:
+
+  * :guilabel:`Clamped to terrain`: takes elevation directly from the terrain
+    height and ignores any existing Z values in the features. A data-defined
+    :guilabel:`Offset` value from the terrain can also be filled.
+  * :guilabel:`Relative to terrain`: any existing Z values in the features
+    is added to the terrain height. A :guilabel:`Scale` factor followed by
```
Comment (nyalldawson) [inline docs/user_manual/working_with_vector/vector_properties.rst:3275]:
```suggestion
    are added to the terrain height. A :guilabel:`Scale` factor followed by
```

File: docs/user_manual/working_with_vector/vector_properties.rst [hunk 4]
```diff
@@ -3235,6 +3245,58 @@ More information on variables usage in the General Tools
...
+    This option is not available for 2D geometry layers.
+  * :guilabel:`Absolute`: ignores the terrain height and directly takes Z values
+    from the features for the elevation. A :guilabel:`Scale` factor followed
+    by a data-defined :guilabel:`Offset` can be used to adjust the elevation.
+    For 2D geometry layers (with no Z values), a data-defined
+    :guilabel:`Base height` can instead be set.
+* |unchecked| :guilabel:`Enable extrusion`: you can set a :guilabel:`Height`
+  to control how high features vertically extend above their base.
+  This is convenient to represent 2D geometry layers, e.g. polygon building
+  footprints layers.
```
Comment (nyalldawson) [inline docs/user_manual/working_with_vector/vector_properties.rst:3286]:
```suggestion 
  This is convenient to indicate that a 2D geometry layers, e.g. a polygon
  building footprints layer actually represents 3D objects.
```

File: docs/user_manual/working_with_vector/vector_properties.rst (hunk 2 above)
Comment (nyalldawson) [reply docs/user_manual/working_with_vector/vector_properties.rst:3295]: Maybe this bad diagram helps explain:

image

the terrain is shown in orange. With "centroid" clamping (blue feature) the line terrain height is only sampled at one place, the centroid, and so the start of the line will "float" above the surface and dig into the surface at the end.

With vertex clamping (pink feature) the terrain height is taken at every vertex, so the extruded line will sit above the terrain at the start and end.

Does that help?

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**